_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,417 @@
+    l <- c(l, loss$item())
+  }
+
+  cat(sprintf("Loss at epoch %d: %3f\n", epoch, mean(l)))
+}
+```
+
+    Loss at epoch 1: 1.795564
+    Loss at epoch 2: 1.540063
+    Loss at epoch 3: 1.495343
+    Loss at epoch 4: 1.461649
+    Loss at epoch 5: 1.446628
+
+Although there is a lot more that *could* be done -- calculate metrics or evaluate performance on a validation set, for example -- the above is a typical (if simple) template for a `torch` training loop.
+
+The optimizer-related idioms in particular
+
+```{r}
+optimizer$zero_grad()
+# ...
+loss$backward()
+# ...
+optimizer$step()
+```
+
+you'll keep encountering over and over.
+
+Finally, let's evaluate model performance on the test set.
+
+## Evaluation
+
+Putting a model in `eval` mode tells `torch` *not* to calculate gradients and perform backprop during the operations that follow:
+
+```{r}
+model$eval()
+```
+
+We iterate over the test set, keeping track of losses and accuracies obtained on the batches.
+
+```{r}
+test_losses <- c()
+total <- 0
+correct <- 0
+
+for (b in enumerate(test_dl)) {
+  output <- model(b[[1]]$to(device = "cuda"))
+  labels <- b[[2]]$to(device = "cuda")
+  loss <- nnf_cross_entropy(output, labels)
+  test_losses <- c(test_losses, loss$item())
+  # torch_max returns a list, with position 1 containing the values 
+  # and position 2 containing the respective indices
+  predicted <- torch_max(output$data(), dim = 2)[[2]]
+  total <- total + labels$size(1)
+  # add number of correct classifications in this batch to the aggregate
+  correct <- correct + (predicted == labels)$sum()$item()
+}
+
+mean(test_losses)
+```
+
+    [1] 1.53784480643349
+
+Here is mean accuracy, computed as proportion of correct classifications:
+
+```{r}
+test_accuracy <-  correct/total
+test_accuracy
+```
+
+    [1] 0.9449
+
+That's it for our first `torch` example. Where to from here?
+
+## Learn
+
+To learn more, check out our vignettes on the `torch` [website](https://mlverse.github.io/torch). To begin, you may want to check out these in particular:
+
+-   "Getting started" series: Build a simple neural network from scratch, starting from [low-level tensor manipulation](https://mlverse.github.io/tohttps://mlverse.github.io/torch/articles/getting-started/tensors.html) and gradually adding in higher-level features like [automatic differentiation](https://mlverse.github.io/torch/articles/getting-started/tensors-and-autograd.html) and [network modules](https://mlverse.github.io/torch/articles/getting-started/nn.html).
+-   More on tensors: [Tensor creation](https://mlverse.github.io/torch/articles/tensor-creation.html) and [indexing](https://mlverse.github.io/torch/articles/indexing.html)
+-   Backpropagation in `torch`: [autograd](https://mlverse.github.io/torch/articles/using-autograd.html)
+
+If you have questions, or run into problems, please feel free to ask on [GitHub](https://github.com/mlverse/torch) or on the [RStudio community forum](https://community.rstudio.com/).
```

Comment:
```suggestion
If you have questions, or run into problems, please feel free to ask on [GitHub](https://github.com/mlverse/torch) or on the [RStudio community forum](https://community.rstudio.com/c/ml/15).
```

Reply:
^^ links them directly to the machine learning and modelling category

Reply:
awesome